### PR TITLE
Fixed #10070 -- Added support for pyformat style parameters on SQLite.

### PR DESCRIPTION
**Pull request by ryancheley**
Fix to allow named parameters on raw sql queries with sqlite

### CODE REVIEW

File: django/db/backends/sqlite3/features.py [hunk 1]
```diff
@@ -18,7 +18,7 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     atomic_transactions = False
     can_rollback_ddl = True
     can_create_inline_fk = False
-    supports_paramstyle_pyformat = False
+    supports_paramstyle_pyformat = True
```
Comment (charettes) [inline django/db/backends/sqlite3/features.py:21]: Ah I assume this is why tests were passing on SQLite as well?

**Review by charettes — COMMENTED** [review]
Thanks for patch Ryan!

We'll want to adjust the docs added in d097417025e71286ad5bbde6e0a79caacabbbd64 to remove the mention of SQLite exception

https://github.com/django/django/blob/970f61fefb148284fb2af63b5cc844279254111a/docs/ref/databases.txt#L822-L828

File: django/db/backends/sqlite3/features.py (hunk 1 above)
Comment (ryancheley) [reply django/db/backends/sqlite3/features.py:21]: @charettes Yes! There is a test called [test_pyformat_params](https://github.com/django/django/blob/0dd29209091280ccf34e07c9468746c396b7778e/tests/raw_query/tests.py#L231) with a `skipUnlessDBFeature` which was causing the test to be skipped. 

I’ll also make an update to the docs (once I am able to incorporate the feedback from the issue into the solution 😁)

**Comment by ryancheley** [comment]
Looks like some tests for `pyformat` are failing and *perhaps* the initial change doesn't indicate an actual support of pyformat. I'll be looking into that next

**Comment by felixxm** [comment]
> Looks like some tests for `pyformat` are failing and _perhaps_ the initial change doesn't indicate an actual support of pyformat. I'll be looking into that next

It works, but the same is needed for `executemany()`.

**Comment by ryancheley** [comment]
@charettes @felixxm I’m not sure why the tests are failing here. 

When I run it locally the tests pass (macOS 12.6 Monterey virtual environment with Python 3.10.2) so I’m kind of at a loss as to what to do ( I did forget to run the linters on my code locally but have done that now …. but I don’t want to push up another change until I can try and figure out why the tests aren’t passing in CI)

Also, the code as submitted work (locally at least) but I think it can be refactored though I’m not **really** sure how. 

Any input you have on it would be welcome!

**Comment by charettes** [comment]
@ryancheley the postgres test failure is a fluke that isn't related to your changes. It's safe to ignore it and focus on the linting failures.

File: django/db/backends/sqlite3/base.py [hunk 2]
```diff
@@ -365,10 +365,22 @@ class SQLiteCursorWrapper(Database.Cursor):
...
             return Database.Cursor.execute(self, query)
+        if hasattr(params, "keys"):
+            args = {k: ":%s" % k for k in params}
+            query = query % args
         query = self.convert_query(query)
         return Database.Cursor.execute(self, query, params)
 
     def executemany(self, query, param_list):
+        # breakpoint()
+        param_list = [p for p in param_list]
```
Comment (felixxm) [inline django/db/backends/sqlite3/base.py:376]: Maybe
```suggestion
        if param_list:
            param_list = list(param_list)
```

**Review by ngnpope — CHANGES_REQUESTED** [review]
There are a couple of issues with the approach:
- There is duplication which should be pushed down into `.convert_query()`.
- Converting `%(value)s` to `:value` in this way will also convert `%%s` to `%s` which will then be incorrectly converted to `?` in `.convert_query()`. We should not perform both conversions[^1] and a test should be added to ensure this doesn't happen.
- Converting `param_list` to a `list` in `.executemany()` will saddle us with poor performance. We many have been passed a generator and the current approach will force all of the parameters to be materialized. The approach needed requires that we can peek at the first item in the generator without consuming the lot.

I'll provide a PR to your branch so that you can see what I mean.

[^1]: We're either converting `format` paramstyle to `qmark` or the `pyformat` paramstyle to `named`. Note that the `sqlite3` [documentation](https://docs.python.org/3/library/sqlite3.html#sqlite3.paramstyle) only mentions `qmark` and `numeric`, but the SQLite [documentation](https://www.sqlite.org/lang_expr.html#parameters) indicates that `named` is also supported. See also the [relevant section](https://peps.python.org/pep-0249/#paramstyle) in PEP 249. I'll try to get the `sqlite3` documentation updated.

**Comment by ngnpope** [comment]
> I'll provide a PR to your branch so that you can see what I mean.

See https://github.com/ryancheley/django/pull/1.

**Comment by ryancheley** [comment]
@felixxm is there anything left for me to do with this PR? If so, please let me know. 😁

**Comment by felixxm** [comment]
@ryancheley Thanks :+1:

I simplified comments.

**Pull request merged**